mne/preprocessing/_csd.py

```diff
@@ -0,0 +1,190 @@
+# Copyright 2003-2010 Jürgen Kayser <rjk23@columbia.edu>
+# Copyright 2017 Federico Raimondo <federaimondo@gmail.com> and
+#                Denis A. Engemann <dengemann@gmail.com>
+#
+#
+# The original CSD Toolbox can be find at
+# http://psychophysiology.cpmc.columbia.edu/Software/CSDtoolbox/
+
+# Authors: Denis A. Engeman <denis.engemann@gmail.com>
+#          Alex Rockhill <aprockhill206@gmail.com>
+#
+# License: Relicensed under BSD (3-clause) and adapted with
+#          permission from authors of original GPL code
+
+import numpy as np
+
+from scipy.linalg import inv
+
+from mne import pick_types
+from mne.utils import _validate_type, _ensure_int
+from mne.io import BaseRaw
+from mne.io.constants import FIFF
+from mne.epochs import BaseEpochs
+from mne.evoked import Evoked
+from mne.bem import fit_sphere_to_headshape
+from mne.channels.interpolation import _calc_g, _calc_h
+
+
+def _prepare_G(G, lambda2):
+    G.flat[::len(G) + 1] += lambda2
+    # compute the CSD
+    Gi = inv(G)
+
+    TC = Gi.sum(0)
+    sgi = np.sum(TC)  # compute sum total
+
+    return Gi, TC, sgi
+
+
+def _compute_csd(G_precomputed, H, radius):
+    """Compute the CSD."""
+    n_channels = H.shape[0]
+    data = np.eye(n_channels)
+    mu = data.mean(0)
+    Z = data - mu
+
+    Gi, TC, sgi = G_precomputed
+
+    Cp2 = np.dot(Gi, Z)
+    c02 = np.sum(Cp2, axis=0) / sgi
+    C2 = Cp2 - np.dot(TC[:, np.newaxis], c02[np.newaxis, :])
+    X = np.dot(C2.T, H).T / radius ** 2
+    return X
+
+
+def compute_current_source_density(inst, sphere='auto', lambda2=1e-5,
+                                   stiffness=4, n_legendre_terms=50,
+                                   copy=True):
+    """Get the current source density (CSD) transformation.
+
+    Transformation based on spherical spline surface Laplacian [1]_ [2]_ [3]_.
+
+    Parameters
+    ----------
+    inst : instance of Raw, Epochs or Evoked
+        The data to be transformed.
+    sphere : array-like, shape (4,) | str
+        The sphere, head-model of the form (x, y, z, r) where x, y, z
+        is the center of the sphere and r is the radius in meters.
+        Can also be "auto" to use a digitization-based fit.
+    lambda2 : float
+        Regularization parameter, produces smoothnes. Defaults to 1e-5.
+    stiffness : float
+        Stiffness of the spline. Also referred to as `m`.
+    n_legendre_terms : int
+        Number of Legendre terms to evaluate.
+    copy : bool
+        Whether to overwrite instance data or create a copy.
+
+    Returns
+    -------
+    inst_csd : instance of Epochs or Evoked
+        The transformed data. Output type will match input type.
+
+    Notes
+    -----
+    This function applies an average reference to the data.
```

Comment:
this suggests it modifiies the input instance. It's not the case unless copy is False